### PR TITLE
Improve filtering of candidate binding invocations in config binder gen

### PR DESCRIPTION
**Pull request by layomia**
Address https://github.com/dotnet/runtime/pull/89587#discussion_r1289058376.

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Helpers/Parser/BinderInvocation.cs [hunk 1]
```diff
@@ -1,84 +1,72 @@
...
+                    ContainingType.Name: string containingTypeName,
+                })
             {
                 return false;
             }
 
             return (containingTypeName) switch
             {
                 "ConfigurationBinder" =>
-                    containingNamespaceName is "Microsoft.Extensions.Configuration" &&
```
Comment (eiriktsarpalis) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Helpers/Parser/BinderInvocation.cs:75]: Don't we still need to check the namespaces of the symbols?

File: src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Helpers/Parser/BinderInvocation.cs [hunk 2]
```diff
@@ -1,84 +1,72 @@
...
+using System.Diagnostics;
 using System.Threading;
 using Microsoft.CodeAnalysis.CSharp.Syntax;
 using Microsoft.CodeAnalysis.Operations;
 using Microsoft.CodeAnalysis;
 
 namespace Microsoft.Extensions.Configuration.Binder.SourceGeneration
 {
-    internal sealed record BinderInvocation
+    internal readonly record struct BinderInvocation
```
Comment (eiriktsarpalis) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Helpers/Parser/BinderInvocation.cs:12]: Why does this need to be a struct?

File: src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Helpers/Parser/BinderInvocation.cs [hunk 3]
```diff
@@ -1,84 +1,72 @@
...
+                    Location = invocationSyntax.GetLocation()
+                };
             }
 
-            return new BinderInvocation()
-            {
-                Operation = operation,
-                Location = invocationSyntax.GetLocation()
-            };
+            return default;
```
Comment (eiriktsarpalis) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Helpers/Parser/BinderInvocation.cs:32]: Why not just return `null` as before? Filtering incremental values with `Where` is preferable to having to detect invalid values in the parser layer.

**Review by eiriktsarpalis — COMMENTED** [review]
Structurally the change looks good, but I had a few questions about the details.

File: src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Helpers/Parser/BinderInvocation.cs (hunk 2 above)
Comment (layomia) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Helpers/Parser/BinderInvocation.cs:12]: Looking at this [doc](https://learn.microsoft.com/en-us/dotnet/standard/design-guidelines/choosing-between-class-and-struct), prob better to revert back to a ref type.

File: src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Helpers/Parser/BinderInvocation.cs (hunk 1 above)
Comment (layomia) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Helpers/Parser/BinderInvocation.cs:75]: Thought we might save on the `ToDisplayString()` alloc; I think the chances of a namespace match are high if the other checks are `true`.

Comment (eiriktsarpalis) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Helpers/Parser/BinderInvocation.cs:75]: It's ok to skip those checks if you were doing an early check on the syntax node, but this is the method that checks against the symbol. You still need to check you're in the right namespace somewhere.

Comment (layomia) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Helpers/Parser/BinderInvocation.cs]: Reverting to allocating the display string.

File: src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Helpers/Parser/BinderInvocation.cs [hunk 4]
```diff
@@ -1,74 +1,61 @@
...
-                return true;
-            }
-
-            invocationSyntax = null;
-            return false;
+                Expression: MemberAccessExpressionSyntax
+                {
+                    Name.Identifier.ValueText: string memberName
+                }
+            } && IsCandidateBindingMethodName(memberName);
```
Comment (eiriktsarpalis) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Helpers/Parser/BinderInvocation.cs:33]: I think you might be able to drill down even more on this expression, roughly:
```suggestion
                    Name.Identifier.ValueText: string memberName,
                    Expression: TypeSyntax declaringType
                }
            } && IsCandidateBindingMethodName(declaringType, memberName);
```

File: src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Helpers/Parser/BinderInvocation.cs [hunk 5]
```diff
@@ -1,74 +1,61 @@
...
 
             static bool IsCandidateBindingMethodName(string name) =>
                 IsCandidateMethodName_ConfigurationBinder(name) ||
                 IsCandidateMethodName_OptionsBuilderConfigurationExtensions(name) ||
                 IsValidMethodName_OptionsConfigurationServiceCollectionExtensions(name);
         }
 
-        private static bool IsCandidateInvocation(IInvocationOperation operation)
+        private static bool IsCandidateInvocationOperation(IInvocationOperation operation)
```
Comment (eiriktsarpalis) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Helpers/Parser/BinderInvocation.cs:41]: The predicate acting on symbols should provide a definitive answer, therefore not producing candidates.
```suggestion
        private static bool IsInvocationOperation(IInvocationOperation operation)
```

**Review by eiriktsarpalis — APPROVED** [review]
Thanks Layomi!

File: src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Helpers/Parser/BinderInvocation.cs [hunk 6]
```diff
@@ -1,74 +1,62 @@
...
-                invocationSyntax = syntax;
-                return true;
-            }
-
-            invocationSyntax = null;
-            return false;
+                Expression: MemberAccessExpressionSyntax
+                {
+                    Name.Identifier.ValueText: string memberName,
+                    Expression: NameSyntax,
```
Comment (eiriktsarpalis) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Helpers/Parser/BinderInvocation.cs:32]: Maybe adding a TODO here tracking declaring type name detection.

Comment (layomia) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Helpers/Parser/BinderInvocation.cs:32]: Done; will add to-do.

File: src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Helpers/Parser/BinderInvocation.cs (hunk 4 above)
Comment (layomia) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Helpers/Parser/BinderInvocation.cs]: I've constrained to `NameSyntax`es for now; follow up - https://github.com/dotnet/runtime/issues/90687.

**Comment by layomia** [comment]
`NameSyntax` clause is causing test failures; reverted in https://github.com/dotnet/runtime/pull/90619/commits/339bcb9f8848dcd3a79822fb92ffc48174f70319. Will address in https://github.com/dotnet/runtime/issues/90687.

**Pull request merged**